### PR TITLE
Refactor old tests to match current format

### PR DESCRIPTION
**Pull request by armandojaleo**
Refactor old tests to match current format

### CODE REVIEW

File: test/instruments_previous_test.go [hunk 1]
```diff
@@ -6,96 +6,133 @@ import (
...
 	"github.com/stretchr/testify/assert"
 
+	"github.com/checkout/checkout-sdk-go/common"
 	"github.com/checkout/checkout-sdk-go/errors"
 	"github.com/checkout/checkout-sdk-go/instruments"
 	"github.com/checkout/checkout-sdk-go/instruments/abc"
 	"github.com/checkout/checkout-sdk-go/tokens"
 )
 
 func TestCreateAndGetInstrumentPrevious(t *testing.T) {
```
Comment (martinseco) [inline test/instruments_previous_test.go:16]: This should be different tests. You should test the Create method first and validate the create response, and then you should do another function to test the Get method with a previously created Instrument.

Comment (martinseco) [reply test/instruments_previous_test.go:16]: Like there should be two different functions:
- TestCreateInstrument
- TestGetInstrument

File: test/instruments_previous_test.go [hunk 2]
```diff
@@ -6,96 +6,133 @@ import (
...
-	updateResponse, err := PreviousApi().Instruments.Update(createResponse.Id, updateRequest)
-	assert.Nil(t, err)
-	assert.NotNil(t, updateResponse)
-	assert.Equal(t, instruments.Card, updateResponse.Type)
-	assert.NotEmpty(t, updateResponse.Fingerprint)
+	cases := []struct {
+		name          string
+		responseId    string
+		updateRequest abc.UpdateInstrumentRequest
+		checker       func(*abc.UpdateInstrumentResponse, error)
```
Comment (martinseco) [inline test/instruments_previous_test.go:73]: Here you could do something similar to what we have in `TestUpdateWorkflowAction` where we use a `putChecker` to validate the response from the updating operation and then we have a `getChecker` to validate that the data was correctly modified. This test should be called `TestUpdateInstrument`

File: test/instruments_previous_test.go [hunk 3]
```diff
@@ -6,96 +6,133 @@ import (
...
+	client := PreviousApi().Instruments
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			tc.checker(client.Update(tc.responseId, tc.updateRequest))
+		})
+	}
 }
 
 func TestCreateAndDeleteInstrumentPrevious(t *testing.T) {
```
Comment (martinseco) [inline test/instruments_previous_test.go:104]: Change the name to TestDeleteInstrumentPrevious

File: test/instruments_previous_test.go [hunk 4]
```diff
@@ -112,5 +149,21 @@ func createTokenInstrumentPrevious(t *testing.T, token *tokens.CardTokenResponse
...
+	assert.NotEmpty(t, response.Last4)
+	assert.NotEmpty(t, response.Bin)
+	assert.NotEmpty(t, response.CardType)
+	assert.NotEmpty(t, response.CardCategory)
+	assert.NotEmpty(t, response.ProductId)
+	assert.NotEmpty(t, response.ProductType)
+	assert.NotEmpty(t, response.Customer)
+	assert.NotEmpty(t, response.Customer.Id)
+	assert.NotEmpty(t, response.Customer.Name)
+	assert.NotEmpty(t, response.Customer.Email)
```
Comment (martinseco) [inline test/instruments_previous_test.go:167]: This should all be asserted in the TestCreateInstrument function. Here we should only validate if there was an error in the response and in that case do a t.Fail("with some error message"). You can check the `createWorkflow(t *testing.T) *common.IdResponse` function

File: test/instruments_test.go [hunk 5]
```diff
@@ -14,120 +14,201 @@ import (
 )
 
 func TestCreateAndGetInstrument(t *testing.T) {
```
Comment (martinseco) [inline test/instruments_test.go:16]: Basically the same comments from test/instruments_previous_test apply here.

File: test/instruments_test.go [hunk 6]
```diff
@@ -14,120 +14,201 @@ import (
...
-	assert.NotEmpty(t, getCardInstrumentResponse.ProductId)
-	assert.NotEmpty(t, getCardInstrumentResponse.ProductType)
+
+	cases := []struct {
+		name       string
+		responseId string
+		checker    func(*nas.GetInstrumentResponse, error)
+	}{
+		{
+			name:       "when get a created instrument then return it",
```
Comment (martinseco) [inline test/instruments_test.go:26]: Check the names for the cases in other tests files like customers_test and adapt them so that they are more readable

Comment (martinseco) [reply test/instruments_test.go:26]: This one would be better as "when fetching a valid instrument then return instrument data"

Comment (martinseco) [reply test/instruments_test.go:26]: Please check the names for all the cases

File: test/payments_actions_previous_test.go [hunk 7]
```diff
@@ -5,28 +5,46 @@ import (
...
-		assert.NotEmpty(t, action.ResponseCode)
-		assert.NotEmpty(t, action.ResponseSummary)
-		assert.NotEmpty(t, action.Type)
+	cases := []struct {
+		name         string
+		instrumentId string
+		checker      func(*abc.GetPaymentActionsResponse, error)
+	}{
+		{
+			name:         "when get a payment then return a response",
```
Comment (martinseco) [inline test/payments_actions_previous_test.go:21]: "when payment id is valid then return payment actions"

File: test/payments_actions_previous_test.go [hunk 8]
```diff
@@ -5,28 +5,46 @@ import (
...
+				}
+			},
+		},
 	}
 
+	client := PreviousApi().Payments
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			Wait(time.Duration(3))
```
Comment (martinseco) [inline test/payments_actions_previous_test.go:46]: Maybe use retriable instead of a wait here?

File: test/payments_actions_test.go [hunk 9]
```diff
@@ -5,28 +5,46 @@ import (
 	"time"
 
 	"github.com/stretchr/testify/assert"
+
+	"github.com/checkout/checkout-sdk-go/payments/nas"
 )
 
 func TestGetPaymentActions(t *testing.T) {
```
Comment (martinseco) [inline test/payments_actions_test.go:12]: Same comments as in TestGetPaymentActionsPrevious

File: test/payments_captures_previous_test.go [hunk 10]
```diff
@@ -24,22 +23,40 @@ func TestCaptureCardPaymentPrevious(t *testing.T) {
...
-	assert.NotEmpty(t, response.Links)
-	assert.NotEmpty(t, response.Links["payment"])
+	cases := []struct {
+		name           string
+		paymentId      string
+		captureRequest abc.CaptureRequest
+		checker        func(*payments.CaptureResponse, error)
+	}{
+		{
+			name:           "when get a capture payment request then return a response",
```
Comment (martinseco) [inline test/payments_captures_previous_test.go:33]: "when request is valid then capture payment"

File: test/payments_captures_previous_test.go [hunk 11]
```diff
@@ -24,22 +23,40 @@ func TestCaptureCardPaymentPrevious(t *testing.T) {
...
+				assert.NotEmpty(t, response.Links["payment"])
+			},
+		},
+	}
+
+	client := PreviousApi().Payments
 
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			Wait(time.Duration(3))
```
Comment (martinseco) [inline test/payments_captures_previous_test.go:51]: Retriable maybe?

File: test/payments_captures_previous_test.go [hunk 12]
```diff
@@ -50,14 +67,37 @@ func TestCaptureCardPaymentIdempotentlyPrevious(t *testing.T) {
...
-	assert.NotNil(t, response2)
+	cases := []struct {
+		name           string
+		paymentId      string
+		captureRequest abc.CaptureRequest
+		idempotencyKey string
+		checker        func(*payments.CaptureResponse, error)
+	}{
+		{
+			name:           "when get a capture payment with idempotencyKey request then return a response",
```
Comment (martinseco) [inline test/payments_captures_previous_test.go:78]: "when request is valid then capture payment idempotently"

File: test/payments_captures_test.go [hunk 13]
```diff
@@ -24,35 +23,53 @@ func TestCaptureCardPayment(t *testing.T) {
...
+				assert.Nil(t, err)
+				assert.NotNil(t, response)
+				assert.NotEmpty(t, response.Reference)
+				assert.NotEmpty(t, response.ActionId)
+				assert.NotEmpty(t, response.Links)
+				assert.NotEmpty(t, response.Links["payment"])
+
+				Wait(time.Duration(3))
+
+				payment, err := DefaultApi().Payments.GetPaymentDetails(paymentResponse.Id)
```
Comment (martinseco) [inline test/payments_captures_test.go:46]: Maybe do something like the getChecker here like in webhooks

File: test/payments_payouts_previous_test.go [hunk 14]
```diff
@@ -32,20 +30,39 @@ func TestRequestPayoutPrevious(t *testing.T) {
...
-
-	//Links
-	assert.NotEmpty(t, response.Links["self"])
+	cases := []struct {
+		name          string
+		payoutRequest abc.PayoutRequest
+		checker       func(*abc.PaymentResponse, error)
+	}{
+		{
+			name:          "when get a payout request then return a response",
```
Comment (martinseco) [inline test/payments_payouts_previous_test.go:39]: Change this name

File: test/payments_captures_test.go [hunk 15]
```diff
@@ -63,14 +80,37 @@ func TestCaptureCardPaymentIdempotently(t *testing.T) {
...
+		{
+			name:           "when get a capture payment with idempotencyKey request then return a response",
+			paymentId:      paymentResponse.Id,
+			captureRequest: captureRequest,
+			idempotencyKey: idempotencyKey,
+			checker: func(response *payments.CaptureResponse, err error) {
+				assert.Nil(t, err)
+				assert.NotNil(t, response)
+
+				response2, err := DefaultApi().Payments.CapturePayment(paymentResponse.Id, captureRequest, &idempotencyKey)
```
Comment (martinseco) [inline test/payments_captures_test.go:99]: Same as in TestCaptureCardPaymentIdempotentlyPrevious

File: test/payments_payouts_previous_test.go [hunk 16]
```diff
@@ -1,17 +1,15 @@
 package test
 
 import (
-	"testing"
-
 	"github.com/stretchr/testify/assert"
+	"testing"
```
Comment (martinseco) [inline test/payments_payouts_previous_test.go:5]: Order this import

File: test/payments_refunds_previous_test.go [hunk 17]
```diff
@@ -11,30 +11,46 @@ import (
...
-	assert.NotEmpty(t, response.Links["payment"])
 
+	cases := []struct {
+		name          string
+		paymentId     string
+		refundRequest payments.RefundRequest
+		checker       func(*payments.RefundResponse, error)
+	}{
+		{
+			name:          "when get a refund payment then return a response",
```
Comment (martinseco) [inline test/payments_refunds_previous_test.go:27]: Change this name

File: test/payments_refunds_previous_test.go [hunk 18]
```diff
@@ -43,14 +59,37 @@ func TestRefundCardPaymentIdempotentlyPrevious(t *testing.T) {
...
+		{
+			name:           "when get a refund payment with idempotencyKey request then return a response",
+			paymentId:      paymentResponse.Id,
+			refundRequest:  refundRequest,
+			idempotencyKey: idempotencyKey,
+			checker: func(response *payments.RefundResponse, err error) {
+				assert.Nil(t, err)
+				assert.NotNil(t, response)
+
+				response2, err := PreviousApi().Payments.RefundPayment(paymentResponse.Id, &refundRequest, &idempotencyKey)
```
Comment (martinseco) [inline test/payments_refunds_previous_test.go:78]: Same as in captures. This should be another test case if it is necessary

File: test/payments_refunds_test.go [hunk 19]
```diff
@@ -11,41 +11,57 @@ import (
...
+			refundRequest: refundRequest,
+			checker: func(response *payments.RefundResponse, err error) {
+				assert.Nil(t, err)
+				assert.NotNil(t, response)
+				assert.NotEmpty(t, response.Reference)
+				assert.NotEmpty(t, response.ActionId)
+				assert.NotEmpty(t, response.Links)
+				assert.NotEmpty(t, response.Links["payment"])
+
+				payment, err := DefaultApi().Payments.GetPaymentDetails(paymentResponse.Id)
```
Comment (martinseco) [inline test/payments_refunds_test.go:38]: Use getChecker func as in workflows

File: test/payments_refunds_test.go [hunk 20]
```diff
@@ -54,14 +70,37 @@ func TestRefundCardPaymentIdempotently(t *testing.T) {
...
+		{
+			name:           "when get a refund payment with idempotencyKey request then return a response",
+			paymentId:      paymentResponse.Id,
+			refundRequest:  refundRequest,
+			idempotencyKey: idempotencyKey,
+			checker: func(response *payments.RefundResponse, err error) {
+				assert.Nil(t, err)
+				assert.NotNil(t, response)
+
+				response2, err := DefaultApi().Payments.RefundPayment(paymentResponse.Id, &refundRequest, &idempotencyKey)
```
Comment (martinseco) [inline test/payments_refunds_test.go:89]: Use getChecker func as in workflows

File: test/payments_request_previous_test.go [hunk 21]
```diff
@@ -12,7 +12,6 @@ import (
 )
 
 func TestRequestCardPaymentPrevious(t *testing.T) {
```
Comment (martinseco) [inline test/payments_request_previous_test.go:14]: What about this one? It's in the legacy format

File: test/payments_request_previous_test.go [hunk 22]
```diff
@@ -44,24 +41,41 @@ func TestRequestPaymentListPrevious(t *testing.T) {
...
-	assert.NotNil(t, paymentListResponse.Data[0].Id)
-	assert.NotNil(t, paymentListResponse.Data[0].RequestedOn)
-	assert.NotNil(t, paymentListResponse.Data[0].Source)
+	cases := []struct {
+		name         string
+		queryRequest payments.QueryRequest
+		checker      func(*abc.GetPaymentListResponse, error)
+	}{
+		{
+			name:         "when get a payment list request then return a response",
```
Comment (martinseco) [inline test/payments_request_previous_test.go:50]: "when payment is valid then return payment list"

File: test/payments_request_previous_test.go [hunk 23]
```diff
@@ -44,24 +41,41 @@ func TestRequestPaymentListPrevious(t *testing.T) {
...
+	client := PreviousApi().Payments
 
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			tc.checker(client.RequestPaymentList(tc.queryRequest))
+		})
+	}
 }
 
 func TestMakeCardVerificationPrevious(t *testing.T) {
```
Comment (martinseco) [inline test/payments_request_previous_test.go:78]: This should be TestRequestPayment

File: test/payments_request_previous_test.go [hunk 24]
```diff
@@ -79,214 +93,84 @@ func TestMakeCardVerificationPrevious(t *testing.T) {
...
+				assert.NotEmpty(t, processing.RetrievalReferenceNumber)
+
+				//Risk
+				assert.False(t, response.Risk.Flagged)
+
+				//Links
+				assert.NotEmpty(t, response.Links["self"])
+				assert.NotEmpty(t, response.Links["actions"])
+				assert.Empty(t, response.Links["capture"])
+				assert.Empty(t, response.Links["void"])
```
Comment (martinseco) [inline test/payments_request_previous_test.go:159]: I don't know if this ones are necessary, but maybe we could create new functions to do this assertions and make this cases more readable. Like: assertCustomer, assertProccessing, etc

File: test/payments_request_previous_test.go [hunk 25]
```diff
@@ -306,20 +190,39 @@ func TestMakePaymentsIdempotentlyPrevious(t *testing.T) {
...
+	}{
+		{
+			name:           "when get a request payment with idempotencyKey then return a response",
+			paymentRequest: paymentRequest,
+			idempotencyKey: idempotencyKey,
+			checker: func(response *abc.PaymentResponse, err error) {
+				assert.Nil(t, err)
+				assert.NotNil(t, response)
+
+				response2, err := PreviousApi().Payments.RequestPayment(paymentRequest, &idempotencyKey)
```
Comment (martinseco) [inline test/payments_request_previous_test.go:207]: Maybe another test case?

File: test/payments_request_previous_test.go [hunk 26]
```diff
@@ -79,214 +93,84 @@ func TestMakeCardVerificationPrevious(t *testing.T) {
...
+	client := PreviousApi().Payments
 
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			tc.checker(client.RequestPayment(tc.paymentRequest, nil))
+		})
+	}
 }
 
 func TestMakePaymentsIdempotentlyPrevious(t *testing.T) {
```
Comment (martinseco) [inline test/payments_request_previous_test.go:173]: TestRequestPaymentIdempotently

File: test/payments_request_previous_test.go [hunk 27]
```diff
@@ -344,8 +247,7 @@ func makeCardPaymentPrevious(t *testing.T, shouldCapture bool, amount int64) *ab
 	return response
 }
 
-func make3dsCardPaymentPrevious(t *testing.T, attemptN3d bool) *abc.PaymentResponse {
-
+func TestMake3dsCardPaymentPrevious(t *testing.T) {
```
Comment (martinseco) [inline test/payments_request_previous_test.go:250]: This one should be another case in TestRequestPayment

File: test/payments_request_test.go [hunk 28]
```diff
@@ -31,11 +31,9 @@ func TestRequestCardPayment(t *testing.T) {
 	assert.Nil(t, paymentResponse.ThreeDs)
 
 	paymentCommonAssertions(t, paymentResponse)
-
 }
 
 func TestMakeCardVerification(t *testing.T) {
```
Comment (martinseco) [inline test/payments_request_test.go:36]: TestRequestPayment

File: test/payments_request_test.go [hunk 29]
```diff
@@ -130,174 +145,39 @@ func TestRequestPaymentList(t *testing.T) {
...
-
-	//Risk
-	assert.False(t, paymentResponse.Risk.Flagged)
+	cases := []struct {
+		name         string
+		queryRequest payments.QueryRequest
+		checker      func(*nas.GetPaymentListResponse, error)
+	}{
+		{
+			name:         "when get a payment list request then return a response",
```
Comment (martinseco) [inline test/payments_request_test.go:154]: Same with the name here

File: test/payments_request_test.go [hunk 30]
```diff
@@ -317,16 +197,36 @@ func TestMakePaymentsIdempotently(t *testing.T) {
...
+	}{
+		{
+			name:           "when get a request payment with idempotencyKey then return a response",
+			paymentRequest: paymentRequest,
+			idempotencyKey: idempotencyKey,
+			checker: func(response *nas.PaymentResponse, err error) {
+				assert.Nil(t, err)
+				assert.NotNil(t, response)
+
+				response2, err := DefaultApi().Payments.RequestPayment(paymentRequest, &idempotencyKey)
```
Comment (martinseco) [inline test/payments_request_test.go:214]: Another test case

File: test/payments_request_test.go [hunk 31]
```diff
@@ -378,8 +278,7 @@ func makeCardPayment(t *testing.T, shouldCapture bool, amount int64) *nas.Paymen
 	return response
 }
 
-func make3dsCardPayment(t *testing.T, attemptN3d bool) *nas.PaymentResponse {
-
+func TestMake3dsCardPayment(t *testing.T) {
```
Comment (martinseco) [inline test/payments_request_test.go:281]: This should be a case in TestRequestPayments

File: test/payments_request_test.go [hunk 32]
```diff
@@ -412,27 +298,152 @@ func make3dsCardPayment(t *testing.T, attemptN3d bool) *nas.PaymentResponse {
...
+	client := DefaultApi().Payments
 
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			tc.checker(client.RequestPayment(tc.paymentRequest, nil))
+		})
+	}
+}
+
+func TestMakeCardTokenPayment(t *testing.T) {
```
Comment (martinseco) [inline test/payments_request_test.go:446]: Same here, it's another case in TestPaymentRequest

File: test/payments_voids_previous_test.go [hunk 33]
```diff
@@ -11,39 +11,77 @@ import (
...
-	assert.NotEmpty(t, response.Links)
-	assert.NotEmpty(t, response.Links["payment"])
+	cases := []struct {
+		name        string
+		paymentId   string
+		voidRequest payments.VoidRequest
+		checker     func(*payments.VoidResponse, error)
+	}{
+		{
+			name:        "when get a voids payment then return a response",
```
Comment (martinseco) [inline test/payments_voids_previous_test.go:27]: "when payment exists then void payment"

File: test/payments_voids_test.go [hunk 34]
```diff
@@ -11,52 +11,90 @@ import (
...
-	assert.Equal(t, int64(0), payment.Balances.AvailableToRefund)
-	assert.Equal(t, int64(0), payment.Balances.AvailableToVoid)
+	cases := []struct {
+		name        string
+		paymentId   string
+		voidRequest payments.VoidRequest
+		checker     func(*payments.VoidResponse, error)
+	}{
+		{
+			name:        "when get a voids payment then return a response",
```
Comment (martinseco) [inline test/payments_voids_test.go:27]: Change the names of this tests

**Review by martinseco — CHANGES_REQUESTED** [review]
Great work here! Left some comments. Please try to remove the Waits when possible and use Retriable

File: test/instruments_previous_test.go (hunk 1 above)
Comment (armandojaleo) [reply test/instruments_previous_test.go:16]: But for get an instrument, we need to create one, and this creation method is called from more than one test, this is the reason.

File: test/instruments_previous_test.go (hunk 3 above)
Comment (armandojaleo) [reply test/instruments_previous_test.go:104]: 👍

File: test/instruments_previous_test.go (hunk 4 above)
Comment (armandojaleo) [reply test/instruments_previous_test.go:167]: 👍

File: test/instruments_previous_test.go (hunk 2 above)
Comment (armandojaleo) [reply test/instruments_previous_test.go:73]: 👍

File: test/instruments_test.go (hunk 5 above)
Comment (armandojaleo) [reply test/instruments_test.go:16]: 👍

File: test/instruments_test.go (hunk 6 above)
Comment (armandojaleo) [reply test/instruments_test.go:26]: 👍

File: test/payments_refunds_previous_test.go (hunk 18 above)
Comment (armandojaleo) [reply test/payments_refunds_previous_test.go:78]: We need to check `assert.Equal(t, response.ActionId, response2.ActionId)`

**Review by martinseco — APPROVED** [review]

**Pull request merged**